stdlib/src/utils/span.mojo

```diff
@@ -114,6 +118,28 @@ struct Span[
     # Life cycle methods
     # ===------------------------------------------------------------------===#
 
+    @doc_private
+    @implicit
+    @always_inline("nodebug")
+    fn __init__(out self: Self.mut, other: Self.immut):
+        """Implicitly cast the immmutable origin of self to a mutable one.
+
+        Args:
+            other: The Span to cast.
+        """
+        self = rebind[__type_of(self)](other)
```

Comment:
I didn't put much thought into it either. Now that you mention it it seems like a possible foot-gun.

The only use case I can think of is when trying to work around the compiler's restriction to having 2 mutable references to the same origin inside a function. An example (for `Pointer`) would be sorting algorithms's inner functions where they'd need 2 references to be compared and then swapped, and if they pointed to the same memory location the compiler would enforce that they should be immutable refs. Currently the sort module just uses `UnsafePointer` to avoid such restrictions.

I'm still not sure though whether we should offer this constructor even as an explicit one, it seems like encouraging function signature "contract breaking".

The more I think about it the less I like the idea, I'll revert the changes to leave only `Immutable` -> `Mutable` and if we find a solid use case we could always implement the counter direction.

Reply:
I agree that we could conceivable want an "unsafe" way to case an immutable origin to a mutable origin (we have that already in a less ergonomic form in `_lit_mut_cast`), but I agree its a special case, and we could leave it out of this PR for the time being.

Thanks for removing those for the time being Martin — I'll give this another look now! 🙂